benchbuild/experiment.py

```diff
@@ -26,30 +26,36 @@ class HelloExperiment(Experiment):
 """
 import collections
 import copy
+import typing as tp
 import uuid
 from abc import abstractmethod
-from typing import List
 
 import attr
 
+import benchbuild.utils.actions as actns
 from benchbuild.settings import CFG
-from benchbuild.utils.actions import (Any, Clean, CleanExtra, Compile,
-                                      Containerize, Echo, MakeBuildDir,
-                                      RequireAll, Run)
 from benchbuild.utils.requirements import Requirement
 
+from . import source
+from .project import Project
+
+Actions = tp.MutableSequence[actns.Step]
+ProjectT = tp.Type[Project]
+Projects = tp.List[ProjectT]
+
 
 class ExperimentRegistry(type):
     """Registry for benchbuild experiments."""
 
     experiments = {}
 
-    def __init__(cls, name, bases, _dict):
+    def __new__(mcs: tp.Type[tp.Any], name: str, bases: tp.Tuple[type, ...],
+                attrs: tp.Dict[str, tp.Any]) -> tp.Any:
         """Register a project in the registry."""
-        super(ExperimentRegistry, cls).__init__(name, bases, _dict)
-
-        if cls.NAME is not None:
+        cls = super(ExperimentRegistry, mcs).__new__(mcs, name, bases, attrs)
+        if bases and hasattr(cls, 'NAME'):
             ExperimentRegistry.experiments[cls.NAME] = cls
+        return cls
```

Comment:
__new__ is when a new object is being created. We are using a metaclass, so this is being called when a class becomes visible to benchbuild.

If we use __init__ or __init_subclass__, we would get the completed class handed to us. Now, while __init__ works just fine (did in the past, there is nothing wrong with it), annotating it with correct type information is a nightmare.
You have no idea about type-information of your descendants in the metaclass, so the assumed type is Any. No type Any has NAME/DOMAIN/GROUP... class attributes.

This above is way cleaner. During class construction we do not have to make any assumption about the descendant class of this metaclass. Before construction, we simply check the attributes that are about to be added to the constructed class.
However, we need to do check the complete type hierachy of the descendant class for any required attributes, because they can be defined in a superclass.

 The only thing I don't like about this, is simply that we have to filter abstract/partial experiments properly (which is not a bad thing). Otherwise, they end up in the experiment registry and cause a mess when doing a run over all experiments (mostly a feature used during testing, but hey...).